Issue 11: Add page on using a custom :fact-type-fn to improve perform…

WilliamParker opened the pull request:
…ance

WilliamParker commented:
This is a draft that I'll probably modify a bit later, at least proofreading it, but I wanted to see if anyone has thoughts on the content.

mrrodriguez commented on _docs/fact_type_fn_perf.md:
```diff
@@ -0,0 +1,82 @@
+	 )
+{% endhighlight %}	 
+
+Where the elipses represent rules that are missing but follow the pattern laid out.  In this case, the id from each IdHolder would be extracted and compared for equality with 1000 diifferent strings, but since these conditions are mutually exclusive this shouldn't be necessary.  The reason for this comparison is that every fact is evaluated against every condition for which its fact type makes it ...
```
Instead of the phrase (multiple places here) "fact type makes it eligible", isn't it more typical to say "fact type matches"?

WilliamParker replied:
I was trying to avoid implying that the fact type had to be equal to the type in the condition since it can also inherit from that type.  Perhaps I was going to far in the direction of being abstract though.

mrrodriguez replied:
I suppose that is fair.  It was a minor suggestion anyways.

mrrodriguez commented on _docs/fact_type_fn_perf.md:
```diff
@@ -0,0 +1,82 @@
+	 )
+(mk-session :fact-type-fn (fn [fact]
+                                                 (cond 
+												        (instance? IdHolder fact) [IdHolder (:id fact)]
```
I'm assuming this indentation is rendering correctly outside of this GitHub view of the markdown...

WilliamParker replied:
I'll check on that.  I'm not sure if you have Jekyll set up on your machine; if you do it is pretty easy to start a local server so you can view the rendered pages in your browser.

````
cd clara-site-directory
bundle install   (for Ruby gems)
jekyll serve
````

mrrodriguez replied:
Yeah I have Jekyll set up. Just wasn't rendering your PR with it. It just printed the indentation weird in this GitHub view.

WilliamParker replied:
The indentation is indeed wrong when I render this in my browser; I'll fix it.

mrrodriguez commented on _docs/fact_type_fn_perf.md:
```diff
@@ -0,0 +1,82 @@
+														:else (type fact))))
+{% endhighlight %}	 
+
+Clara sessions can be roughly though of as containing a map from fact types to conditions that facts of that type are eligible to meet.  This custom :fact-type-fn function will determine the type of each IdHolder fact for rule processing services and then look that type up in this map, rather than essentially iterating over all of the map values to find which ones match.
```
Isn't the typical Clara terminology "constraint" here.  Perhaps the sentence would read better as:
"a map from fact types to constraints on the facts that match the associated fact type"

WilliamParker replied:
I think "condition" is correct.. perhaps we're talking about different things here?

https://github.com/cerner/clara-rules/blob/0.13.0/src/main/clojure/clara/rules/schema.clj#L28

If your point is that some constraints on conditions will be evaluated in nodes deeper than the AlphaNode, I agree; I just wasn't looking to get into that level of detail here since it seems irrelevant to the matter at hand.

mrrodriguez replied:
From that schema (and from terminology used throughout the library), a condition is a fact type together with a collection of constraints.

So in this paragraph you are splitting out the constituents of a condition.  The fact type is the key and associated with it is a (possibly empty) collection of constraints to apply to those facts who "matched on (or the eligible phrasing you say above)" the fact type.

This is a minor detail though and it probably doesn't matter to most people.

WilliamParker replied:
Ah, I see what you're saying now.  I think "condition" is probably better than "constraint" in this context because when the AlphaNode is activated with a fact that fact is then (possibly transitively) evaluated for all of the constraints in the condition.  There are, as you say, more precise ways we could parse this out but that seems like unnecessary complication of the subject of this page.  I'm a bit torn between saying "condition" and "rule" actually but at the moment I'm still leaning toward "condition" between the two.

mrrodriguez replied:
That's fine.  I'd definitely favor "condition" over "rule".  I think "rule" is overused and represents a bigger conceptual structure.

mrrodriguez commented:
The only issue I have here is that the `:ancestors-fn` isn't really discussed.  In the example you use, you sort of implicitly handled the `:ancestors-fn` job in the `:fact-type-fn` by performing an `instance?` check on the type and effectively causing all subtypes of the type to be reduced to the single ancestor type.

In the general case, I think being sure to not break the `:ancestors-fn` is important though, so probably worth mentioning more?

WilliamParker commented:
This probably does need to be more explicit about the consequences for matching ancestors of using a custom :fact-type-fn while also not setting the :ancestors-fn.  My immediate thought is to write a sentence or two about it and link back the to page on custom :fact-type-fn and :ancestors-fn functions at http://www.clara-rules.org/docs/fact_type_customization/

WilliamParker committed:
Issue 11: Add documenting how to use a custom :fact-type-fn to improve performance

WilliamParker commented:
I've updated the PR in response to the comments.  In particular, I added a sentence discussing the interaction of the :fact-type-fn and :ancestors-fn, fixed some indentation problems, and clarified/corrected my phrasing in a few places.

EthanEChristian commented on _docs/fact_type_fn_perf.md:
```diff
@@ -0,0 +1,75 @@
+	  ;; do something
+	  )
+(defrule rule3
+    [[IdHolder "AAA")]
```
nit: ) -> ]

WilliamParker replied:
I fixed this and updated the PR.

WilliamParker commented:
I think everything has been addressed satisfactorily with my latest updates to the PR and the changes weren't particularly major.  I'm going to go ahead and merge this, but we can make any desired improvements easily enough later.
